scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala

```diff
@@ -149,11 +150,19 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
     *
     * @return A random Float, and the next Randomizer to use.
     */
-  def nextFloat: (Float, Randomizer) = { // Uses same algorithm as ScalaCheck for this one
-    val (s, rs) = chooseInt(0, 1)
-    val (e, re) = chooseInt(0, 0xfe)
-    val (m, rm) = chooseInt(0, 0x7fffff)
-    (java.lang.Float.intBitsToFloat((s << 31) | (e << 23) | m), rm)
+  def nextFloat: (Float, Randomizer) = {
+    // The space of possible Floats
+    // 2 to the power of 32, which is covered by Int.MinValue to Int.MaxValue
+    // - (2 to the power of 23 (23 is the number of mantissa bits))
+    // + (3 for the two infinities and one NaN)
+
+    // scala> Int.MaxValue - math.pow(2, 23).toInt + 2
+    // res13: Int = 2139095042
+    val (x, r) = chooseInt(Int.MinValue, 2139095042) 
+
+    // Pick one lucky number to play the lotto with:
+    if (x == 999) (Float.NaN, r)
+    else r.nextExtRealFloatValue
```

Comment:
Mostly at an aesthetic level, I'm a little unhappy about the inefficiency of the new algorithm -- especially when you factor in what `nextExtRealFloatValue` does, we're generating two random numbers solely for the fantasmagorically small chance that they will result in one of your magic numbers.  It's minor, but it's wasted cycles in virtually all cases.

Can/should we compose these operations somehow?  This feels like there might be a better factoring...

Reply:
I'll think about this when I look at our chooseInt algos, which is coming up.